#!/bin/bash 
# uses file generated in simulateGenotypes.R to generate VCF


start=$(date +%s)

#TODO Change path
#PATH_OUT=~/Emily/Data/test_NGSNGS/simulated
PATH_OUT=/media/rna/WHAKAHAO/Emily/Data/simulatedData/vcfSimulation

LABEL=Try2

VCFFILE=$PATH_OUT/simulatedVCF_${LABEL}.vcf 
VCF_R_DATA=$PATH_OUT/data4VCF_${LABEL}.txt
META=data4VCF_${LABEL}_meta.txt



echo "##fileformat=VCFv4.2" > $VCFFILE
echo "##fileDate=$(date +'%Y%m%d')" >> $VCFFILE
echo "##source=$0" >> $VCFFILE
echo '##FILTER=<ID=PASS,Description="All filters passed">' >> $VCFFILE

# adds reference from "meta" file, second part omits name of meta entry
echo "##reference=$(grep referenceGenome $PATH_OUT/$META | awk '{$1=""; print substr($0,2)}')" >> $VCFFILE

#adds contig information
CHROMOSOME=$(grep chromosomeName $PATH_OUT/$META | awk '{$1=""; print substr($0,2)}')
LENGTH=$(grep chromosomeLength $PATH_OUT/$META | awk '{$1=""; print substr($0,2)}')
echo -e "##contig=<ID=$CHROMOSOME,length=$LENGTH>" >> $VCFFILE
echo '##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">' >> $VCFFILE

# second part omits name of meta entry
SAMPLES=$(grep sampleNames ${PATH_OUT}/${META} | awk '{$1=""; print substr($0,2)}' | tr ' ' '\t')
echo -e "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\t$SAMPLES" >> $VCFFILE

# add variants from dat4VCF (generated by simulateGenotypes.R)
# format of dat4VCF:  
# POS	refAllele	altAllele	Indv1	Indv2 ...
awk -v chr=${CHROMOSOME} 'NR > 1 {printf chr"\t%s\t.\t%s\t%s\t100\tPASS\t.\tGT", $1, $2, $3; for (i=4; i<=NF; i++) printf "\t%s", $i; printf "\n"}' ${VCF_R_DATA} >> ${VCFFILE}

#less VCFFILE

end=$(date +%s)
runtime=$((end - start))
echo "Program took $runtime seconds"
echo "generateVCF.sh\t${runtime}" >> $PATH_OUT/../timingSimulation_${LABEL}.txt